src/compositor/output.c

```diff
@@ -868,12 +868,52 @@ wlc_output_get_resolution(wlc_handle output)
    return get(convert_from_wlc_handle(output, "output"), offsetof(struct wlc_output, resolution));
 }
 
+bool
+wlc_output_set_scale_ptr(struct wlc_output *output, int32_t scale)
+{
+   if (!output)
+      return false;
+
+   assert(scale);
+   wlc_log(WLC_LOG_WARN, "Set scale of output %" PRIuWLC "to %u", convert_to_wlc_handle(output), scale);
+   output->information.scale=scale;
+   return true;
+}
 WLC_API void
 wlc_output_set_resolution(wlc_handle output, const struct wlc_size *resolution)
 {
    wlc_output_set_resolution_ptr(convert_from_wlc_handle(output, "output"), resolution);
 }
 
+WLC_API void
+wlc_output_set_scale(wlc_handle output, int32_t scale)
+{
+   wlc_output_set_scale_ptr(convert_from_wlc_handle(output, "output"), scale);
+}
+WLC_API int32_t
+wlc_output_get_scale(wlc_handle output)
+{
+   struct wlc_output *o = convert_from_wlc_handle(output, "output");
+   assert(o);
+   return o->information.scale;
+}
+
+WLC_API bool
```

Comment:
non-integral scaling can be achieved by combination of scale and wlc_output_set_resolution. Non integral scaling for surfaces is not supported by wayland protocol. Also supporting that would be really PITA when you consider buffer sizes could be non integral. Thus it's better to set nearest integral scale to output and then compensate with rendering resolution (not neccessarily modeline).